### ISSUE TITLE
Cannot login with android app

### ISSUE THREAD

**Issue by ishioni** [nextcloud/user_oidc#537]
I've setup nextcloud 25 in conjunction with authentik. It works well in the browser and somewhat well with the desktop app. However, the mobile android app refuses to even access the login screen, and i can see 

```
172.16.133.235 - - [21/Nov/2022:01:18:51 +0000] "GET /status.php HTTP/1.1" 200 181 "-" "Mozilla/5.0 (Android) Nextcloud-android/3.22.3" "10.1.2.52"
172.16.133.235 - - [21/Nov/2022:01:18:51 +0000] "HEAD /remote.php/dav HTTP/1.1" 401 0 "-" "Mozilla/5.0 (Android) Nextcloud-android/3.22.3" "10.1.2.52" 
```

Logs on the client side:

```
11-21 08:36:37.665 29341  7535 I ServerNameIndicator: SSLSocket implementation: com.google.android.gms.org.conscrypt.Java8FileDescriptorSocket
11-21 08:36:37.665 29341  7535 I ServerNameIndicator: SNI done, hostname: <domain>
11-21 08:36:37.811 29341  7535 W HttpMethodBase: Going to buffer response body of large or unknown size. Using getResponseBodyAsStream instead is recommended.
11-21 08:36:37.812 29341  7535 I GetStatusRemoteOperation: Connection check at https://<domain>: Operation finished with HTTP status code -1 (success)
11-21 08:36:37.812 29341  7535 D GetServerInfoOperation: Trying empty authorization to detect authentication method
11-21 08:36:37.815 29341  7535 D OwnCloudClient #1: REQUEST HEAD /remote.php/dav
11-21 08:36:37.925 29341  7535 D RemoteOperationResult: RemoteOperationResult has processed UNHANDLED_HTTP_CODE: 405 Method Not Allowed
11-21 08:36:37.925 29341  7535 D ExistenceCheckRemoteOperation: Existence check for https://<domain>/remote.php/dav/files/null// targeting for  existence finished with HTTP status 405(FAIL)
11-21 08:36:37.926 29341  7535 D DetectAuthenticationMethodOperation: Authentication method found: UNKNOWN
11-21 08:36:37.926 29341  7535 D OperationsService: Called 1 listeners
11-21 08:36:37.926 29341  7535 D OperationsService: Stopping after command with id 1
```

When trying to login. Is it possible to use the mobile app with this?

**Comment by jacksgt** [comment]
I just wanted to chime in and report that I'm seeing the same behavior with `user_oidc` `v1.2.1` and Nextcloud `24.0.5`:

HTTP log:
```
10.42.0.18 - - [26/Nov/2022:17:18:40 +0000] "GET /status.php HTTP/1.1" 200 1681 "-" "Mozilla/5.0 (Android) Nextcloud-android/3.22.3"
10.42.0.18 - - [26/Nov/2022:17:18:40 +0000] "HEAD /remote.php/dav HTTP/1.1" 401 1236 "-" "Mozilla/5.0 (Android) Nextcloud-android/3.22.3"
```

Nextcloud log:
```
{"reqId":"DDjoBjkOTB4NKKBnvYVq","level":3,"time":"2022-11-26T17:18:41+00:00","remoteAddr":"10.42.0.3","user":"--","app":"user_oidc","method":"GET","url":"/ocs/v2.php/cloud/user?format=json","message":"Impossible to decode OIDC token:Wrong number of segments","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":{"app":"user_oidc"}}
{"reqId":"DDjoBjkOTB4NKKBnvYVq","level":3,"time":"2022-11-26T17:18:41+00:00","remoteAddr":"10.42.0.3","user":"--","app":"user_oidc","method":"GET","url":"/ocs/v2.php/cloud/user?format=json","message":"Could not find unique token validation","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":{"app":"user_oidc"}}
```

(This is with the `Check Bearer token on API requests` option enabled)

Let me know if it's necessary to provide any other logs for debugging / troubleshooting.

**Comment by jacksgt** [comment]
Interestingly enough, this does *not* occur when I have the LDAP user backend enabled (in parallel to `user_oidc`).
Is it possible that `user_ldap` implements some logic for intercepting "app passwords" (which is missing from `user_oidc`)?

**Comment by juliushaertl** [comment]
> Impossible to decode OIDC token:Wrong number of segments

This looks a bit like maybe it is trying to take the app password provided by the client as an JWT token. Could you maybe disable the bearer token validation and paste the nextcloud logs (ideally at debug level) and access logs with another attempt of that? 

> Interestingly enough, this does not occur when I have the LDAP user backend enabled (in parallel to user_oidc).
> Is it possible that user_ldap implements some logic for intercepting "app passwords" (which is missing from user_oidc)?

Seems strange, I'd not be aware of any special handling in that regard in user_ldap. 

cc @julien-nc I don't have a instance right away that I can test on Android, but maybe you have one so you can check as I don't remember any issue when testing with keycloak OIDC last time.

**Comment by jacksgt** [comment]
As requested, here are the `DEBUG` server logs (*without* the bearer token validation for API requests):

```
{"reqId":"yUL4YdRf17oRb2dv6Slf","level":0,"time":"2022-11-28T19:42:15+00:00","remoteAddr":"10.42.0.3","user":"--","app":"no app in context","method":"HEAD","url":"/remote.php/dav","message":"OC_App::registerLogIn() is deprecated, please register your alternative login option using the registerAlternativeLogin() on the RegistrationContext in your Application class implementing the OCP\\Authentication\\IAlternativeLogin interface","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":[]}
{"reqId":"yUL4YdRf17oRb2dv6Slf","level":0,"time":"2022-11-28T19:42:15+00:00","remoteAddr":"10.42.0.3","user":"--","app":"webdav","method":"HEAD","url":"/remote.php/dav","message":"No public access to this resource., No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured, No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","exception":{"Exception":"Sabre\\DAV\\Exception\\NotAuthenticated","Message":"No public access to this resource., No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured, No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured","Code":0,"Trace":[{"file":"/var/www/html/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"beforeMethod","class":"Sabre\\DAV\\Auth\\Plugin","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":456,"function":"emit","class":"Sabre\\DAV\\Server","type":"->","args":["beforeMethod:HEAD",[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":253,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":321,"function":"start","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/lib/Server.php","line":358,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/remote.php","line":166,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Auth/Plugin.php","Line":152,"CustomMessage":"--"}}
{"reqId":"ZFImXpFRavb7Z7JyvSIo","level":0,"time":"2022-11-28T19:42:16+00:00","remoteAddr":"10.42.0.3","user":"--","app":"no app in context","method":"GET","url":"/ocs/v2.php/cloud/user?format=json","message":"OC_App::registerLogIn() is deprecated, please register your alternative login option using the registerAlternativeLogin() on the RegistrationContext in your Application class implementing the OCP\\Authentication\\IAlternativeLogin interface","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":[]}
{"reqId":"ZFImXpFRavb7Z7JyvSIo","level":3,"time":"2022-11-28T19:42:16+00:00","remoteAddr":"10.42.0.3","user":"--","app":"user_oidc","method":"GET","url":"/ocs/v2.php/cloud/user?format=json","message":"Could not find unique token validation","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":{"app":"user_oidc"}}
{"reqId":"mFMBngD8uvGOdgkna6ku","level":0,"time":"2022-11-28T19:42:41+00:00","remoteAddr":"10.42.0.3","user":"--","app":"no app in context","method":"HEAD","url":"/remote.php/dav/files/jh//","message":"OC_App::registerLogIn() is deprecated, please register your alternative login option using the registerAlternativeLogin() on the RegistrationContext in your Application class implementing the OCP\\Authentication\\IAlternativeLogin interface","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":[]}
{"reqId":"mFMBngD8uvGOdgkna6ku","level":3,"time":"2022-11-28T19:42:41+00:00","remoteAddr":"10.42.0.3","user":"--","app":"user_oidc","method":"HEAD","url":"/remote.php/dav/files/jh//","message":"Could not find unique token validation","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":{"app":"user_oidc"}}
{"reqId":"JmKEjyZ0C81QDStytxtB","level":0,"time":"2022-11-28T19:43:06+00:00","remoteAddr":"10.42.0.3","user":"--","app":"no app in context","method":"GET","url":"/ocs/v2.php/cloud/capabilities?format=json","message":"OC_App::registerLogIn() is deprecated, please register your alternative login option using the registerAlternativeLogin() on the RegistrationContext in your Application class implementing the OCP\\Authentication\\IAlternativeLogin interface","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":[]}
{"reqId":"JmKEjyZ0C81QDStytxtB","level":3,"time":"2022-11-28T19:43:06+00:00","remoteAddr":"10.42.0.3","user":"--","app":"user_oidc","method":"GET","url":"/ocs/v2.php/cloud/capabilities?format=json","message":"Could not find unique token validation","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.22.3","version":"24.0.5.1","data":{"app":"user_oidc"}}
```

All I'm doing on the client-side is opening the Nextcloud app, clicking on "Login" and scanning the QR app password generated from the Nextcloud web interface.

> Seems strange, I'd not be aware of any special handling in that regard in user_ldap.

Maybe it's the other way around? Is `user_oidc` intercepting the request too early and therefore doesn't let Nextcloud handle the app password authentication properly?
(*Just a random idea*)

**Comment by ishioni** [comment]
@juliushaertl any news on this?

**Comment by jacksgt** [comment]
> @jacksgt Can you describe in your blog how you're setting up authentik with oidc+ldap so that this works?

At the moment it only incidentally works because I have user_oidc + user_ldap enabled (see my comment above https://github.com/nextcloud/user_oidc/issues/537#issuecomment-1328109926 ).
It's unclear what exactly causes this combination to work.
Clearly that's not a real fix, though :-)

**Comment by ishioni** [comment]
It would be a fix enough for me :D I'm having problems configuring authentik for it, so a small how-to would be great. And once this is figured out i could just drop ldap

**Comment by ishioni** [comment]
@jacksgt @juliushaertl I have figured it out! 
#1 nginx-ingress was replacing the response code of the container from 401 to 500 when using custom error pages
#2 Turns out that when nextcloud serves a 401 it does so with a couple of cookies set. The mobile app expects those to proceed. When using custom error pages, those cookies are obviously missing

tl;dr: Don't use custom error pages with nextcloud. After disabling them, oidc works like a charm with both the desktop and mobile apps

**Comment by juliushaertl** [comment]
Might be worth to report this to the Android app to also handle 500 errors more gracefully then

**Comment by jacksgt** [comment]
Hi @ishioni I'm happy you managed to get it working on your side.

However, I'm using Traefik (not Nginx) which doesn't do any custom error pages, therefore I don't think the underlying issue is solved.
